### PR TITLE
`fatxpool`: proper handling of priorities when mempool is full

### PR DESCRIPTION
**Pull request by michalkucharczyk**
Higher-priority transactions can now replace lower-priority transactions even when the internal _tx_mem_pool_ is full.

**Notes for reviewers:**
- The _tx_mem_pool_ now maintains information about transaction priority. Although _tx_mem_pool_ itself is stateless, transaction priority is updated after submission to the view. An alternative approach could involve validating transactions at the `at` block, but this is computationally expensive. To avoid additional validation overhead, I opted to use the priority obtained from runtime during submission to the view. This is the rationale behind introducing the `SubmitOutcome` struct, which synchronously communicates transaction priority from the view to the pool. This results in a very brief window during which the transaction priority remains unknown - those transaction are not taken into consideration while dropping takes place. In the future, if needed, we could update transaction priority using view revalidation results to keep this information fully up-to-date (as priority of transaction may change with chain-state evolution).
- When _tx_mem_pool_ becomes full (an event anticipated to be rare), transaction priority must be known to perform priority-based removal. In such cases, the most recent block known is utilized for validation. I think that speculative submission to the view and re-using the priority from this submission would be an unnecessary complication. 
- Once the priority is determined, lower-priority transactions whose cumulative size meets or exceeds the size of the new transaction are collected to ensure the pool size limit is not exceeded.
- Transaction removed from _tx_mem_pool_ , also needs to be removed from all the views with appropriate event (which is done by `remove_transaction_subtree`). To ensure complete removal, the  `PendingTxReplacement` struct was re-factored to more generic `PendingPreInsertTask` (introduced in #6405) which covers removal and submssion of transaction in the view which may be potentially created in the background. This is to ensure that removed transaction will not re-enter to the newly created view. 
- `submit_local` implementation was also improved to properly handle priorities in case when mempool is full. Some missing tests for this method were also added.

Closes: #5809

### CODE REVIEW

File: substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs [hunk 1]
```diff
@@ -596,18 +694,25 @@ where
...
-			let xt_hash = self.api.hash_and_length(&replacement.xt).0;
-			futures.push(self.replace_transaction_in_view(
-				view.clone(),
-				replacement.source.clone(),
-				replacement.xt.clone(),
-				xt_hash,
-				replacement.watched,
-			));
+		for replacement in self.pending_txs_tasks.read().values() {
+			match replacement.action {
```
Comment (skunert) [inline substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs:698]: Are we protected against double execution of these actions?
I am imagining a scenario where we have [1] and this code here executing in parallel.
In [1] we already have processed then insertion of a `pending_txs_task`, but its not marked as processed yet. Then the codepath here could advance and process the freshly inserted task before it is marked processed, so it will not be removed at the end of `apply_pending_tx_replacements`.

Next time we arive here we will again process the same pending task and then remove it.

[1]: https://github.com/paritytech/polkadot-sdk/blob/2221d7a4ae0d897636ba465046d4bf94cd46e862/substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs#L834-L834

File: substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs [hunk 2]
```diff
@@ -575,8 +669,12 @@ where
 		replaced: ExtrinsicHash<ChainApi>,
 		watched: bool,
 	) {
-		if let Entry::Vacant(entry) = self.pending_txs_replacements.write().entry(replaced) {
-			entry.insert(PendingTxReplacement::new(xt.clone(), source.clone(), watched));
+		if let Entry::Vacant(entry) = self.pending_txs_tasks.write().entry(replaced) {
```
Comment (skunert) [inline substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs:672]: I find it a bit surprising that this whole chain of `replace_xy` is not really replacing, but seems to only submit. If I understand correctly this is
because it happens as a result of a usurped event which means the tx has been removed already?

Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs:672]: Lower layers of txpool (`graph`) does not provide means for explicitly replacing transaction.
Transaction can be simply submitted - and if it replaces any other transaction the event will be triggered.

So basically, what we want to do when a single view reports replacement (_usurpation?_) is to implement the same replacement in every other view kept in the view_store and also from the mempool. So for every view we [check](https://github.com/paritytech/polkadot-sdk/blob/2221d7a4ae0d897636ba465046d4bf94cd46e862/substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs#L784-L793) if replaced transaction is also there, and yes, we replace it.

We want to do it because we don't want the lower prio transaction to stay somewhere in transaction pool.

Basically we could ignore replacing transaction in the rest of the views - the probability of including the replaced transaction is little, but IMO we should cover all edge cases as forking maybe unpredictable, and txpool shall make no assumption on the dynamics of how view will be used.

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 3]
```diff
@@ -671,15 +693,21 @@ where
 		Ok(mempool_results
 				.into_iter()
 				.map(|result| {
-					result.and_then(|insertion| {
+					result
+						.map_err(Into::into)
+						.and_then(|insertion| {
 						submission_results
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: `zip` will not work here. but I added some comments to explain the logic. Hope it makes the code more comprehensible.

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 4]
```diff
@@ -279,27 +326,124 @@ where
...
+		// that the sum of their bytes is ge then size of new tx. Otherwise we could overflow size
+		// limits. Naive way to do it - rev-sort by priority and eat the tail.
+
+		// reverse (oldest, lowest prio last)
+		sorted.sort_by(|(_, a), (_, b)| match b.priority().cmp(&a.priority()) {
+			Ordering::Equal => match (a.source.timestamp, b.source.timestamp) {
+				(Some(a), Some(b)) => b.cmp(&a),
+				_ => Ordering::Equal,
+			},
+			ordering @ _ => ordering,
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: added comments in 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 5]
```diff
@@ -462,6 +602,15 @@ where
 		});
 		self.listener.invalidate_transactions(&invalid_hashes);
 	}
+
+	pub(super) fn update_transaction(&self, outcome: &ViewStoreSubmitOutcome<ChainApi>) {
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 6]
```diff
@@ -1263,6 +1312,69 @@ where
...
+			.verify_one(
+				best_view.at.hash,
+				best_view.at.number,
+				TimedTransactionSource::from_transaction_source(source, false),
+				xt.clone(),
+				crate::graph::CheckBannedBeforeVerify::Yes,
+			)
+			.await;
+
+		if let Some(priority) = validated_tx.priority() {
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 7]
```diff
@@ -1263,6 +1312,69 @@ where
...
+
+				self.view_store.remove_transaction_subtree(
+					*worst_hash,
+					|listener, removed_tx_hash| {
+						listener.limits_enforced(&removed_tx_hash);
+					},
+				);
+			}
+
+			// todo: add to pending_replacements - make sure it will not sneak back via cloned view
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: nope, removed in 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/view.rs [hunk 8]
```diff
@@ -460,4 +460,19 @@ where
 		const IGNORE_BANNED: bool = false;
 		self.pool.validated_pool().check_is_known(tx_hash, IGNORE_BANNED).is_err()
 	}
+
+	/// Removes the whole transaction subtree from the inner pool.
+	///
+	/// Intended to be called when removal is a result of replacement. Provided `replaced_with`
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/view.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs [hunk 9]
```diff
@@ -690,4 +795,59 @@ where
 		};
 		let _results = futures::future::join_all(submit_futures).await;
 	}
+
+	/// Removes a transaction subtree from the every view in the view_store, starting from the given
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 10]
```diff
@@ -1109,7 +1152,13 @@ where
 			.await
 			.into_iter()
 			.zip(hashes)
-			.map(|(result, tx_hash)| result.or_else(|_| Err(tx_hash)))
+			.map(|(result, tx_hash)| {
+				//todo: we may need a bool flag here indicating if we need to update priority
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 11]
```diff
@@ -1263,6 +1312,69 @@ where
...
+	///
+	/// This asynchronous function verifies the new transaction against the most recent view. If a
+	/// transaction with a lower priority exists in the transaction pool, it is replaced with the
+	/// new transaction.
+	///
+	/// If no lower-priority transaction is found, the function returns an error indicating the
+	/// transaction was dropped immediately.
+	async fn attempt_transaction_replacement(
+		&self,
+		_: Block::Hash,
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: caca2e1e81

File: substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs [hunk 12]
```diff
@@ -204,6 +206,10 @@ where
 						views
 					);
 					views.remove(&key);
+					//todo: merge heads up warning!
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/dropped_watcher.rs]: it means to be extra careful when merging with master :)
cleaned in: b86ef05398

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 13]
```diff
@@ -650,9 +654,27 @@ where
...
 
 		if view_store.is_empty() {
-			return Ok(mempool_results.into_iter().map(|r| r.map(|r| r.hash)).collect::<Vec<_>>())
+			return Ok(mempool_results
+				.into_iter()
+				.map(|r| r.map(|r| r.hash).map_err(Into::into))
+				.collect::<Vec<_>>())
 		}
 
+		//todo: review + test maybe?
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs]: 5d0283e

File: substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs (hunk 1 above)
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs:698]: The goal here is to avoid having unwanted transactions in newly cloned view which is being inserted. This whole cloning / building / updating / inserting is executed in context of maintain task. Here we execute tasks for the view that was newly created, and was not in the view_store when the tasks were scheduled.

The `processed` flag indicates that task was accomplished for every view in the view_store. Meaning that newly cloned view (which is not yet inserted) may contain undesired transactions, and tasks need to be executed.

What you say could happen. But it means that new (second) cloned view can be created while tasks are still not completed. Meaning that for this new view, the tasks also need to be executed.

Does it make sense?

Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs:698]: One more thought:

> Next time we arive here we will again process the same pending task and then remove it.

This is expected. Please note that processing will take place for another view (which was possibly cloned when not all tasks were yet processed for every view in view_store).

**Review by skunert — APPROVED** [review]

**Review by iulianbarbu — COMMENTED** [review]
Part 1, will follow up with a closer look at the new tests.

File: substrate/client/transaction-pool/tests/fatp_prios.rs [hunk 14]
```diff
@@ -247,3 +248,257 @@ fn fatp_prio_watcher_future_lower_prio_gets_dropped_from_all_views() {
...
+	let xt1_watcher = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt1.clone())).unwrap();
+	let xt2_watcher = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt2.clone())).unwrap();
+	let xt3_watcher = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt3.clone())).unwrap();
+
+	assert_ready_iterator!(header01.hash(), pool, [xt3, xt0, xt1, xt2]);
+	assert_pool_status!(header01.hash(), &pool, 4, 0);
+	assert_eq!(pool.mempool_len().1, 4);
+
+	// let header02 = api.push_block_with_parent(header01.hash(), vec![], true);
+	// block_on(pool.maintain(new_best_block_event(&pool, Some(header01.hash()), header02.hash())));
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/tests/fatp_prios.rs:359]:
```suggestion
```

File: substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs [hunk 15]
```diff
@@ -690,4 +795,58 @@ where
 		};
 		let _results = futures::future::join_all(submit_futures).await;
 	}
+
+	/// Removes a transaction subtree from every view in the view_store, starting from the given
+	/// transaction hash.
+	///
+	/// This function traverses the dependency graph of transactions and removes the specified
+	/// transaction along with all its descendant transactions from every view.
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/view_store.rs:803]: Transactions may depend on each other which results in (dependency) graph of transactions.
In this case _descendant_ of `tx0` means all transactions that depend on `tx0`. Either directly by their _requires/provides_ tags, or indirectly (through inbetween txs) when there is more items in the dependency chain. You can also re-read this [chapter](https://hackmd.io/@_FY3-hvwQZ6cX_4n8zYUNA/SyT1QuhnA?type=view#What-are-tags) - `Q,R,S,T` are all descendant of `P` (in the last row of this) [picture](https://hackmd.io/_uploads/S1Uiwc_p0.svg).

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 16]
```diff
@@ -617,4 +770,201 @@ mod tx_mem_pool_tests {
 			sc_transaction_pool_api::error::Error::ImmediatelyDropped
 		));
 	}
+
+	#[test]
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs:774]: yeah, good point :)
something I promise to myself when I add new tests - that I will describe them. And I always skipp that part :see_no_evil:. I will try to improve this :)

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 17]
```diff
@@ -279,27 +326,124 @@ where
...
+	/// Attempts to insert a new transaction in the memory pool and drop some worse existing
+	/// transactions.
+	///
+	/// This operation will not overflow the limit of the mempool. It means that cumulative
+	/// size of removed transactions will be equal (or greated) then size of newly inserted
+	/// transaction.
+	///
+	/// Returns a `Result` containing `InsertionInfo` if the new transaction is successfully
+	/// inserted; otherwise, returns an appropriate error indicating the failure.
+	pub(super) fn try_insert_with_dropping(
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs]: Improved in bdf5b2a

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 18]
```diff
@@ -279,27 +326,124 @@ where
...
+		transactions.insert(hash, Arc::from(new_tx));
+		for worst_hash in &to_be_removed {
+			transactions.remove(worst_hash);
+		}
+		debug_assert!(!self.is_limit_exceeded(transactions.len(), self.transactions.bytes()));
+
+		Ok(InsertionInfo::new_with_removed(hash, source, to_be_removed))
+	}
+
+	/// Attempts to insert a new transaction in the memory pool and drop some worse existing
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs]: Done in bdf5b2a.

File: substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs [hunk 19]
```diff
@@ -857,7 +914,13 @@ where
 			.extend_unwatched(TransactionSource::Local, &[xt.clone()])
 			.remove(0)?;
 
-		self.view_store.submit_local(xt).or_else(|_| Ok(xt_hash))
+		self.view_store
+			.submit_local(xt)
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/fork_aware_txpool.rs:921]: good point, I improved it in 8ee3b7b and also added some tests for `submit_local` which were missing in 15739bf.

File: substrate/client/transaction-pool/src/graph/base_pool.rs [hunk 20]
```diff
@@ -453,27 +453,29 @@ impl<Hash: hash::Hash + Member + Serialize, Ex: std::fmt::Debug> BasePool<Hash,
...
+						.map(|worst| {
+							// Here we don't use `TransactionRef`'s ordering implementation because
+							// while it prefers priority like need here, it also prefers older
+							// transactions for inclusion purposes and limit enforcement needs to
+							// prefer newer transactions instead and drop the older ones.
+							match worst.transaction.priority.cmp(&transaction.transaction.priority)
+							{
+								Ordering::Less => worst,
+								Ordering::Equal =>
+									if worst.insertion_id > transaction.insertion_id {
```
Comment (iulianbarbu) [inline substrate/client/transaction-pool/src/graph/base_pool.rs:469]: I see this `insertion_id` of type u64 that comes with the ready txs set of a pool, which gets incremented with each new ready tx imported. It will take a while until this can overflow and flip the logic for a bit (until all the upper end u64 TXs of the ready set are removed), but are there other reasons that makes this harder than having u64::MAX imported TXs that go through the ready set?

Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/graph/base_pool.rs:469]: Not sure I understood a question. Would you please rephrase it?

Comment (iulianbarbu) [reply substrate/client/transaction-pool/src/graph/base_pool.rs:469]: The question is are we concerned with `insertion_id` being wrapped around (U64::MAX -> 0)? In this case an insertion_id = 0 can be considered more recent than a previous insertion_id e.g U64::MAX, so the logic would not stand for this situation.

Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/graph/base_pool.rs:469]: Well, I'd say no. 2^64 is quite a huge number.

The other thing is that this code existed for a while before I started to stick my nose into txpool. So maybe - in future we should update this code to have here a value that is exactly the same across different views (timestamp from `TimedTransactionSource` could be a good candidate).

File: substrate/client/transaction-pool/tests/fatp_prios.rs [hunk 21]
```diff
@@ -247,3 +249,315 @@ fn fatp_prio_watcher_future_lower_prio_gets_dropped_from_all_views() {
...
+
+	api.set_priority(&xt0, 1);
+	api.set_priority(&xt1, 2);
+	api.set_priority(&xt2, 3);
+	api.set_priority(&xt3, 4);
+
+	api.set_priority(&xt4, 5);
+	api.set_priority(&xt5, 6);
+
+	let xt0_watcher = block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt0.clone())).unwrap();
```
Comment (iulianbarbu) [inline substrate/client/transaction-pool/tests/fatp_prios.rs:286]: nit: a bit misguided about the `invalid_hash()`, `Default::default()` can also do it.

```suggestion
	let xt0_watcher = block_on(pool.submit_and_watch(Default::default(), SOURCE, xt0.clone())).unwrap();
```

File: substrate/client/transaction-pool/tests/fatp_prios.rs [hunk 22]
```diff
@@ -247,3 +249,315 @@ fn fatp_prio_watcher_future_lower_prio_gets_dropped_from_all_views() {
...
+	let _xt4_watcher =
+		block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt4.clone())).unwrap();
+	let _xt5_watcher =
+		block_on(pool.submit_and_watch(invalid_hash(), SOURCE, xt5.clone())).unwrap();
+
+	assert_pool_status!(header03.hash(), &pool, 2, 0);
+	assert_eq!(pool.mempool_len().1, 4);
+
+	// let header04 = api.push_block_with_parent(header03.hash(), vec![], true);
+	// block_on(pool.maintain(new_best_block_event(&pool, Some(header03.hash()), header04.hash())));
```
Comment (iulianbarbu) [inline substrate/client/transaction-pool/tests/fatp_prios.rs:315]:
```suggestion
```

**Review by iulianbarbu — APPROVED** [review]

File: substrate/client/transaction-pool/tests/fatp_prios.rs (hunk 21 above)
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/tests/fatp_prios.rs:286]: `deafault` would also do the work, by using `invalid` I wanted to highlight that it is not relevant (param is not used). I also have it in every test, so if we decide to rename, we should do it once in all tests to be consistent.

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs [hunk 23]
```diff
@@ -617,4 +755,201 @@ mod tx_mem_pool_tests {
...
+
+		let results = mempool.extend_unwatched(TransactionSource::External, &xts);
+		assert!(results.iter().all(Result::is_ok));
+		assert_eq!(mempool.bytes(), total_xts_bytes);
+
+		submit_outcomes
+			.into_iter()
+			.for_each(|o| mempool.update_transaction_priority(&o));
+
+		//this one should drop 3 xts (each of size 1129)
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs]: done in b697b0f

File: substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs:759]: I renamed tests in b697b0f, so should a bit easier to reason about the flow.

[the remark this answers is not in the document]
Comment (iulianbarbu) [reply substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs:759]: Yup, this is enough for now, thanks!
